libraries/lib-cloud-audiocom/UploadService.cpp

```diff
@@ -0,0 +1,475 @@
+                  response->readAll<std::string>());
+            }
+            else if (responseCode == 422)
+            {
+               FailPromise(
+                  UploadOperationCompleted::Result::InvalidData,
+                  response->readAll<std::string>());
+            }
+            else
+            {
+               FailPromise(
+                  UploadOperationCompleted::Result::UnexpetedResponse,
+                  response->readAll<std::string>());
+            }
+         });
+   }
+
+   void HandleUploadPolicy(std::string uploadPolicyJSON)
+   {
+      using namespace audacity::network_manager;
+      
+      rapidjson::Document document;
+      document.Parse(uploadPolicyJSON.data(), uploadPolicyJSON.length());
+
+      if (
+         !document.HasMember("url") || !document.HasMember("success") ||
+         !document.HasMember("fail") || !document.HasMember("progress"))
+      {
+         FailPromise(
+            UploadOperationCompleted::Result::UnexpetedResponse,
+            uploadPolicyJSON);
+
+         return;
+      }
+
+      auto form = std::make_unique<MultipartData>();
+
+      if (document.HasMember("fields"))
+      {
+         const auto& fields = document["fields"];
+
+         for (auto it = fields.MemberBegin(); it != fields.MemberEnd(); ++it)
+            form->Add(it->name.GetString(), it->value.GetString());
+      }
+
+      const auto fileField =
+         document.HasMember("field") ? document["field"].GetString() : "file";
+
+      const wxFileName name { mFileName };
+      form->AddFile(fileField, DeduceMimeType(name.GetExt()), name);
+
+      const auto url = document["url"].GetString();
+      
+      mSuccessUrl = document["success"].GetString();
+      mFailureUrl = document["fail"].GetString();
+      mProgressUrl = document["progress"].GetString();
+
+      if (document.HasMember("extra"))
+      {
+         const auto& extra = document["extra"];
+
+         mAudioID = extra["audio"]["id"].GetString();
+         mAudioSlug = extra["audio"]["slug"].GetString();
+
+         if (extra.HasMember("token"))
+            mUploadToken = extra["token"].GetString();
+      }
+
+      const auto encType = document.HasMember("enctype") ?
+                              document["enctype"].GetString() :
+                              "multipart/form-data";
+
+      Request request(url);
+      
+      request.setHeader("Content-Type", encType);
+      request.setHeader("Accept", "application/json");
+
+      // We only lock late and for very short time
+      std::lock_guard<std::mutex> lock(mStatusMutex);
+      if (mAborted)
+         return;
+      
+      auto response =
+         NetworkManager::GetInstance().doPost(request, std::move(form));
+
+      mActiveResponse = response;
+      
+      response->setRequestFinishedCallback(
+         [response, sharedThis = shared_from_this(), this](auto)
+         {
+            HandleS3UploadCompleted(response);
```

Comment:
What does "S3" abbreviate?

Reply:
Well, it kinda means S3 (or Simple Storage Service, but I doubt anyone calls it that). It's a somewhat standard protocol for object storages. 

https://aws.amazon.com/s3/